src/sound/DalSoRiR2.cc

```diff
@@ -0,0 +1,374 @@
+		// WAVE part  0x7E-0x7F
+		switch (port & 0x01) {
+		case 0: // read latch, not supported
+			return 255;
+		case 1: // read wave register
+			return ymf278.peekReg(opl4latch);
+		default:
+			UNREACHABLE;
+		}
+	} else {
+		// FM part  0xC0-0xC7
+		switch (port & 0x03) {
+		case 0: // read status
+		case 2:
+			return ymf262.peekStatus() | readYMF278Status(time);
+		case 1:
+		case 3: // read fm register
+			return ymf262.peekReg(opl3latch);
+		default:
+			UNREACHABLE;
+		}
+	}
+}
+
+void DalSoRiR2::writeIO(word port, byte value, EmuTime::param time)
+{
+	if ((port & 0xFF) < 0xC0) {
+		// WAVE part  0x7E-0x7F
+		if (getNew2()) {
+			switch (port & 0x01) {
+			case 0: // select register
+				ymf278BusyTime = time + WAVE_REG_SELECT_DELAY;
+				opl4latch = value;
+				break;
+			case 1:
+				if ((0x08 <= opl4latch) && (opl4latch <= 0x1F)) {
+					ymf278LoadTime = time + LOAD_DELAY;
+				}
+				if ((3 <= opl4latch) && (opl4latch <= 6)) {
+					// Note: this time is so small that on
+					// MSX you never see BUSY=1 for these
+					// registers. Confirmed on real HW that
+					// also registers 3-5 are faster.
+					ymf278BusyTime = time + MEM_WRITE_DELAY;
+				} else {
+					// For the other registers it is
+					// possible to see BUSY=1, but only
+					// very briefly and only on R800.
+					ymf278BusyTime = time + WAVE_REG_WRITE_DELAY;
+				}
+				if (opl4latch == 0xf8) {
+					ymf262.setMixLevel(value, time);
+				} else if (opl4latch == 0xf9) {
+					ymf278.setMixLevel(value, time);
+				}
+				ymf278.writeReg(opl4latch, value, time);
+				break;
+			default:
+				UNREACHABLE;
+			}
+		} else {
+			// Verified on real YMF278:
+			// Writes are ignored when NEW2=0 (both register select
+			// and register write).
+		}
+	} else {
+		// FM part  0xC0-0xC7
+		switch (port & 0x03) {
+		case 0: // select register bank 0
+			opl3latch = value;
+			ymf278BusyTime = time + FM_REG_SELECT_DELAY;
+			break;
+		case 2: // select register bank 1
+			opl3latch = value | 0x100;
+			ymf278BusyTime = time + FM_REG_SELECT_DELAY;
+			break;
+		case 1:
+		case 3: // write fm register
+			ymf278BusyTime = time + FM_REG_WRITE_DELAY;
+			ymf262.writeReg(opl3latch, value, time);
+			break;
+		default:
+			UNREACHABLE;
+		}
+	}
+}
+
+byte DalSoRiR2::readMem(word addr, EmuTime::param /*time*/)
+{
+	if ((0x3000 <= addr) && (addr < 0x4000)) {
+		// SRAM frame 0
+		return sram[addr - 0x3000];
+	} else if ((0x7000 <= addr) && (addr < 0x8000) && ((regFrame[0] & DISABLE) == 0)) {
+		// SRAM frame 1
+		return sram[(addr - 0x6000) + (regFrame[0] & REG_FRAME_SEGMENT_MASK) * 0x1000];
+	} else if ((0xB000 <= addr) && (addr < 0xC000) && ((regFrame[1] & DISABLE) == 0)) {
+		// SRAM frame 2
+		return sram[(addr - 0x9000) + (regFrame[1] & REG_FRAME_SEGMENT_MASK) * 0x1000];
+	} else {
+		// read selected segment of flash
+		auto flashBank = addr >> 13;
+		auto flashSegmentAddr = addr & 0x3FFF;
+		return flash.read(flashSegmentAddr + regBank[flashBank] * 0x4000);
```

Comment:
Did you forgot to apply `REG_BANK_SEGMENT_MASK` here?
Or maybe better: we should take the actual flash ROM size into account (or does AMDFlash::read() already do that internally?)

In the PCB photos is saw that it uses a SST39SF010A chip. That's a 128kB flash ROM chip.  IIRC the documentation states that the flash size can vary between 128kB and 512kB   (That would be SST39SF010A (1Mbit = 128kB), SST39SF020A (2Mbit = 256kB) or SST39SF040A (4Mbit = 512kB).  512kB indeed requires a mask of 5 bit (32 * 16kB = 512kB). But if we only emulate a 128kB flash ROM, then this address calculation can go out-of-bounds.

I think the easiest fix is to mask on the actual flash rom size. Something like   `(address-calculation) & (flash.size() - 1)`.
EDIT: I checked: AmdFlash::read() does mask internally, ignore this last suggestion.

Reply:
I had it at first, but I realized that the value written in regBank[x] ss always already masked to 5 bits when writing. But now I see I forgot to add that masking there. Will add it.

Reply:
This cannot be done for the frame register, because it has that extra bit. So, this idea I had makes it a bit asymmetric. So I guess I'll indeed just add the mask for the regBank as well.